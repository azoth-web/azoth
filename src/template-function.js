export function compile(ast, renderer) {
    const { expressions, quasis } = ast.quasi;
    const bindings = [];
    return `() => {${compileExpressions(expressions)}${compileBindings(bindings)}
    const __fn = () => {${!hasItems(bindings) 
        ? `
        return ${renderer}().root;` 
        : `
        const { root, nodes } = ${renderer}();
        bind(nodes);
        return root;`}
    };
    return __fn
}`;
}

function compileExpressions(expressions) {
    if(!expressions || expressions.length === 0) return '';
}

const hasItems = arr => arr?.length > 0;

function compileBindings(bindings) {
    if(!hasItems(bindings)) return '';
}

function compileNoBindingsRender() {
    return `    const __fn = () => {
    
    `;}