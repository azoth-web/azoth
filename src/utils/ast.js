import { parse as acornParse } from 'acorn';
import { generate as astringGenerate } from 'astring';

const ACORN_DEFAULTS = {
    ecmaVersion: 'latest',
    sourceType: 'module',
    locations: true,
};

export function parse(source, options = {}) {
    return acornParse(source, { ...ACORN_DEFAULTS, ...options });
}

const ASTRING_DEFAULTS = { 
    indent: '    '
};

export function generate(ast, options = {}) {
    return astringGenerate(ast, { ...ASTRING_DEFAULTS, ...options });
}