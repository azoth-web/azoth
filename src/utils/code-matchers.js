import { diffStringsUnified as diff } from 'jest-diff';
import { expect } from 'vitest';
import { parse, generate } from './ast.js';
import './to-code.js';

const tryParse = (name, code) => {
    try {
        const ast = parse(code);
        return generate(ast);
    }
    catch (err) {
        // eslint-disable-next-line no-console
        console.log('FAILED PARSE:', name, '\nERROR:', err, '\nCODE:\n', code);
        // TODO: why did I not throw this?
        // throw err;
    }
};

function getCode(value, name) {
    if(typeof value === 'function') {
        return { value: value.toCode(), type: 'function' };
    }
    else if(typeof value === 'string') {
        return { 
            value: tryParse(name, value), 
            type: 'string' 
        };
    } 
    return { value: generate(value), type: 'ast' };
}

function codeEqual(received, expected) {
    
    received = getCode(received, 'received');
    expected = getCode(expected, 'expected');
    
    const pass = received.value === expected.value;

    const message = () => {
        const hint = this.utils.matcherHint('toMatchCode',
            `received:${received.type}`, 
            `expected:${expected.type}`, {
                isNot: this.isNot,
                promise: this.promise,
            });

        const diffString = diff(expected.value, received.value, {
            expand: this.expand,
        });  

        return `${hint}\n\n${diffString}\n`;
    };

    return { 
        message, 
        pass 
    };

}

expect.extend({
    toMatchCode: codeEqual,
});
