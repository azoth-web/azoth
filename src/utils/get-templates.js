// test-util

import { simple } from 'acorn-walk';
import { parse } from './ast.js';

// TODO: option for function call
const isAzoth = node => node.tag?.name === '_';

export function getTemplatesAst(code) {
    if(typeof code === 'function') code = code.toCode();
    const ast = parse(code);
    const templates = [];
    simple(ast, {
        TaggedTemplateExpression(node, module, c) {
            // base.TaggedTemplateExpression(node, module, c);
            if(!isAzoth(node)) return;
            templates.push(node);
        },
    });
       
    return templates;
}
