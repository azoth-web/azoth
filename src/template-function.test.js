import { expect, describe, test } from 'vitest';
import './utils/code-matchers.js';
import { compile } from './template-function.js';
import { parse } from './utils/ast.js';
import { simple } from 'acorn-walk';
import { getTemplatesAst } from './utils/get-templates.js';

const _ = () => {};
const isAzoth = node => node.tag?.name === '_';

test('html-only template (no binding)', () => {
    function template() {
        const Greeting = () => _`<span>hello world</span>`;
    }

    const [ast] = getTemplatesAst(template.toCode());
    expect(ast).toMatchSnapshot();

    const code = compile(ast, '__renderer0');

    expect(code).toMatchInlineSnapshot(`
      "() => {
          const __fn = () => {
              return __renderer0().root;
          };
          return __fn
      }"
    `);
});