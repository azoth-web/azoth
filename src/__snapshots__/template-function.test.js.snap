// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`html-only template (no binding) 1`] = `
Node {
  "end": 50,
  "loc": SourceLocation {
    "end": Position {
      "column": 50,
      "line": 1,
    },
    "start": Position {
      "column": 23,
      "line": 1,
    },
  },
  "quasi": Node {
    "end": 50,
    "expressions": [],
    "loc": SourceLocation {
      "end": Position {
        "column": 50,
        "line": 1,
      },
      "start": Position {
        "column": 24,
        "line": 1,
      },
    },
    "quasis": [
      Node {
        "end": 49,
        "loc": SourceLocation {
          "end": Position {
            "column": 49,
            "line": 1,
          },
          "start": Position {
            "column": 25,
            "line": 1,
          },
        },
        "start": 25,
        "tail": true,
        "type": "TemplateElement",
        "value": {
          "cooked": "<span>hello world</span>",
          "raw": "<span>hello world</span>",
        },
      },
    ],
    "start": 24,
    "type": "TemplateLiteral",
  },
  "start": 23,
  "tag": Node {
    "end": 24,
    "loc": SourceLocation {
      "end": Position {
        "column": 24,
        "line": 1,
      },
      "start": Position {
        "column": 23,
        "line": 1,
      },
    },
    "name": "_",
    "start": 23,
    "type": "Identifier",
  },
  "type": "TaggedTemplateExpression",
}
`;
